src/spikeinterface/core/waveform_tools.py

```diff
@@ -397,6 +398,280 @@ def _waveform_extractor_chunk(segment_index, start_frame, end_frame, worker_ctx)
+        return all_waveforms
+    elif mode == "shared_memory":
+        if copy:
+            if shm is not None:
+                # release all sharedmem buffer
+                # empty array have None
+                shm.unlink()
+            return all_waveforms.copy()
+        else:
+            return all_waveforms, wf_array_info
+
+
+def _init_worker_distribute_single_buffer(
+    recording, unit_ids, spikes, wf_array_info, nbefore, nafter, return_scaled, mode, sparsity_mask
+):
+    worker_ctx = {}
+    worker_ctx["recording"] = recording
+    worker_ctx["wf_array_info"] = wf_array_info
+    worker_ctx["spikes"] = spikes
+    worker_ctx["nbefore"] = nbefore
+    worker_ctx["nafter"] = nafter
+    worker_ctx["return_scaled"] = return_scaled
+    worker_ctx["sparsity_mask"] = sparsity_mask
+    worker_ctx["mode"] = mode
+
+    if mode == "memmap":
+        filename = wf_array_info
+        all_waveforms = np.load(str(filename), mmap_mode="r+")
+        worker_ctx["all_waveforms"] = all_waveforms
+    elif mode == "shared_memory":
+        from multiprocessing.shared_memory import SharedMemory
+
+        shm, shm_name, dtype, shape = wf_array_info
+        shm = SharedMemory(shm_name)
+        all_waveforms = np.ndarray(shape=shape, dtype=dtype, buffer=shm.buf)
+        worker_ctx["shm"] = shm
+        worker_ctx["all_waveforms"] = all_waveforms
+
+    # prepare segment slices
+    segment_slices = []
+    for segment_index in range(recording.get_num_segments()):
+        s0 = np.searchsorted(spikes["segment_index"], segment_index)
+        s1 = np.searchsorted(spikes["segment_index"], segment_index + 1)
+        segment_slices.append((s0, s1))
+    worker_ctx["segment_slices"] = segment_slices
+
+    return worker_ctx
+
+
+# used by ChunkRecordingExecutor
+def _worker_distribute_single_buffer(segment_index, start_frame, end_frame, worker_ctx):
+    # recover variables of the worker
+    recording = worker_ctx["recording"]
+    segment_slices = worker_ctx["segment_slices"]
+    spikes = worker_ctx["spikes"]
+    nbefore = worker_ctx["nbefore"]
+    nafter = worker_ctx["nafter"]
+    return_scaled = worker_ctx["return_scaled"]
+    sparsity_mask = worker_ctx["sparsity_mask"]
+    all_waveforms = worker_ctx["all_waveforms"]
+
+    seg_size = recording.get_num_samples(segment_index=segment_index)
+
+    s0, s1 = segment_slices[segment_index]
+    in_seg_spikes = spikes[s0:s1]
+
+    # take only spikes in range [start_frame, end_frame]
+    # this is a slice so no copy!!
+    i0 = np.searchsorted(in_seg_spikes["sample_index"], start_frame)
+    i1 = np.searchsorted(in_seg_spikes["sample_index"], end_frame)
+    if i0 != i1:
+        # protect from spikes on border :  spike_time<0 or spike_time>seg_size
+        # useful only when max_spikes_per_unit is not None
+        # waveform will not be extracted and a zeros will be left in the memmap file
+        while (in_seg_spikes[i0]["sample_index"] - nbefore) < 0 and (i0 != i1):
```

Comment:
You can pre-filter your spikes in the segment before running np.searchsorted for this two conditions to avoid this expensive (and hard to read loop)

Reply:
We need the original index for writting into the buffer.
So this is not possible or would need some extra variable.
This while loop in practice never happen unless the very first spike is very near the border. So this is not expenssive at all.
But this is a bit hard to read I admit it.
Please not not change this it was a very old bug and very hard to track.

Reply:
Which original index? i0? 

Aren't you losing the original variable anyway when you modify it here? I agree that if this never happens maybe doing the vectorization is a bad idea. 

If this is an old bug that was hard to track we should have a test for this. No point on relying on your memory and analytical capabilities if we can pin down the behavior that we want and automatize its check with a test.

Reply:
OK I did a patch as you proposed.